### ISSUE TITLE
The distribution of mlp_up and attention wqkv seems strange in your released files

### ISSUE THREAD

**Issue by milktea888** [FasterDecoding/TEAL#8]
Hi, thank you for your great work, but i seems that I encountered some phenomenon which doesn't match with that in your paper.
The following is the distribution of mlp wup and wgate:
![mlp_layer_0_h15](https://github.com/user-attachments/assets/201d2c08-f949-4f98-adc4-f2276edc9023)
Cause in your paper, you said it typically exhibit Gaussian-like shape.
The data I plot this figure is what you released in the model folder, Llama-2-7b/mlp.

And another question, in the h1(which is the input), the distribution is as flollowing:
![mlp_layer_0](https://github.com/user-attachments/assets/2c52ec11-c075-4dd3-99d4-939ee09278cf)
can I say the input has less 0 than the following modules, like attn_wo or mlp_wdown? Which looks like this:
![mlp_layer_0_h2](https://github.com/user-attachments/assets/0acd622e-6561-4117-a899-44325268b7f5)

**Comment by chromecast56** [comment]
Can you try the following (replace `hist_path`)? Matplotlib tends to have graphical bugs when you try to render too many things (eg. all the histograms) -- the below implementation instead samples 1000 points from the pdf.

```
import matplotlib.pyplot as plt

def interp(x, xp, fp):
    """Custom interpolation function for PyTorch tensors."""
    i = torch.searchsorted(xp, x)
    i = torch.clamp(i, 1, len(xp) - 1)
    
    xp_left = xp[i - 1]
    xp_right = xp[i]
    fp_left = fp[i - 1]
    fp_right = fp[i]
    
    t = (x - xp_left) / (xp_right - xp_left)
    return fp_left + t * (fp_right - fp_left)

class Distribution:
    def __init__(self, bin_centers, counts):
        self.bin_centers = bin_centers
        self.counts = counts
        self.total_count = self.counts.sum()
        self.cumulative_counts = torch.cumsum(self.counts, dim=0)
    
    def pdf(self, x, bandwidth=None):
        if bandwidth is None:
            bandwidth = 0.15 * torch.std(self.bin_centers[1:-1]) * (self.total_count-2)**(-1/5)
        
        bin_centers = self.bin_centers.unsqueeze(1)
        
        if isinstance(x, float) or isinstance(x, int):
            x = torch.tensor([x])
        else:
            x = x.unsqueeze(0)
        
        kernel = torch.exp(-0.5 * ((x - bin_centers) / bandwidth)**2) / (bandwidth * torch.sqrt(torch.tensor(2 * torch.pi)))
        pdf = torch.sum(kernel * self.counts.unsqueeze(1), dim=0) / self.total_count
        
        return pdf
    
    def cdf(self, x):
        return interp(x, self.bin_centers, self.cumulative_counts / self.total_count)

import matplotlib.pyplot as plt
import torch
import numpy as np
from scipy.optimize import curve_fit
from scipy.stats import laplace

def load_histogram(model, layer, module, hist_type):
    hist_path = f"/home/jamesliu/release/TEAL/models/{model}/histograms/layer-{layer}/{module}/histograms.pt"
    hist = torch.load(hist_path)
    bin_centers = hist[f'{hist_type}_centers']
    counts = hist[hist_type]
    return bin_centers, counts

def plot_layer_distributions(model, layer):
    fig, axes = plt.subplots(1, 4, figsize=(30, 5))
    # fig.patch.set_alpha(0)
    
    components = [
        ('mlp', 'h1', "MLP $W_{up,gate}$", 'gaussian'),
        ('mlp', 'h2', "MLP $W_{down}$", 'laplacian'),
        ('self_attn', 'h1', "Attn $W_{Q,K,V}$", 'gaussian'),
        ('self_attn', 'h2', "Attn $W_O$", 'laplacian')
    ]
    
    for idx, (ax, (module, hist_type, label, fit_type)) in enumerate(zip(axes, components)):
        bin_centers, counts = load_histogram(model, layer, module, hist_type)
        dist = Distribution(bin_centers, counts)
        
        x = torch.linspace(bin_centers[1], bin_centers[-2], 1000)
        pdf = dist.pdf(x)
        bin_width = (x[1] - x[0]).item()
        
        # Plot original data
        ax.plot(x.numpy(), pdf.numpy(), linewidth=4)
        ax.bar(x, pdf, width=bin_width, alpha=0.3)
        
        ax.set_title(label, fontsize=40)
        ax.tick_params(axis='both', which='major', labelsize=30)
        
        for spine in ax.spines.values():
            spine.set_linewidth(2.5)

        # ax.patch.set_alpha(0)
    
    fig.text(0.01, 0.55, 'Density', va='center', rotation='vertical', fontsize=40)
    fig.text(0.515, 0.01, 'Activation Value', ha='center', fontsize=40)
    
    plt.tight_layout()
    plt.subplots_adjust(left=0.06, right=0.98, bottom=0.25, top=0.9)
    plt.show()

plot_layer_distributions("Llama-2-7B", 0)

```

![image](https://github.com/user-attachments/assets/e81caf9f-ff66-47ba-bd79-44d1ff5b0d29)

**Comment by milktea888** [comment]
You are so nice that you told me the code! 
And thank you so much, now I can get the correct results.

**Comment by milktea888** [comment]
Maybe could I know what did you mean for the following? That is, I could not understand the y-axis in the results. I know the kernel formula, it's the Gaussian formula, while you multiply a value which in the Gaussian distribution with the "counts"--self.counts.unsqueeze(1). Then you plot pdf in y-axis. Actually, I could not understand why you did so, cause I think if you multiply a value which got by the Gaussian formula, you could always get the Gaussian distribution. So could you just explain the reason you did this? Thank you so much!
'''
kernel = torch.exp(-0.5 * ((x - bin_centers) / bandwidth)**2) / (bandwidth * torch.sqrt(torch.tensor(2 * torch.pi)))
pdf = torch.sum(kernel * self.counts.unsqueeze(1), dim=0) / self.total_count
'''

**Comment by chromecast56** [comment]
Glad it works! Regarding your questions:

1. y axis represents the density in the probability density function (PDF)
2. I use the gaussian kernel to smooth out the distribution for visualization - I think it's fairly standard. Actually, I think you can replace it with a simple moving average and get similar results:

```
import matplotlib.pyplot as plt

def interp(x, xp, fp):
    """Custom interpolation function for PyTorch tensors."""
    i = torch.searchsorted(xp, x)
    i = torch.clamp(i, 1, len(xp) - 1)
    
    xp_left = xp[i - 1]
    xp_right = xp[i]
    fp_left = fp[i - 1]
    fp_right = fp[i]
    
    t = (x - xp_left) / (xp_right - xp_left)
    return fp_left + t * (fp_right - fp_left)

class Distribution:
    def __init__(self, bin_centers, counts):
        self.bin_centers = bin_centers
        self.counts = counts
        self.total_count = self.counts.sum()
        self.cumulative_counts = torch.cumsum(self.counts, dim=0)
    
    # def pdf(self, x, bandwidth=None):
    #     if bandwidth is None:
    #         bandwidth = 0.15 * torch.std(self.bin_centers[1:-1]) * (self.total_count-2)**(-1/5)
        
    #     bin_centers = self.bin_centers.unsqueeze(1)
        
    #     if isinstance(x, float) or isinstance(x, int):
    #         x = torch.tensor([x])
    #     else:
    #         x = x.unsqueeze(0)
        
    #     kernel = torch.exp(-0.5 * ((x - bin_centers) / bandwidth)**2) / (bandwidth * torch.sqrt(torch.tensor(2 * torch.pi)))
    #     pdf = torch.sum(kernel * self.counts.unsqueeze(1), dim=0) / self.total_count
        
    #     return pdf
    
    def pdf(self, x, window_size=50):
        # Convert input to tensor if it's a single number
        if isinstance(x, (float, int)):
            x = torch.tensor([x])
        
        # Compute moving average of counts
        smoothed_counts = torch.nn.functional.avg_pool1d(
            self.counts.unsqueeze(0).unsqueeze(0),
            kernel_size=window_size,
            stride=1,
            padding=window_size // 2
        ).squeeze(0).squeeze(0)
        
        # Interpolate smoothed counts at input x
        pdf = interp(x, self.bin_centers, smoothed_counts)

        return pdf

    def cdf(self, x):
        return interp(x, self.bin_centers, self.cumulative_counts / self.total_count)
```

![image](https://github.com/user-attachments/assets/e9d8a847-12b4-4587-a5cc-8ee8d0d0b72d)

(The big spikes at the ends are because we group the bottom/top 0.1% into single boxes. Also, these distributions isn't normalized)

**Comment by milktea888** [comment]
Yes, the latter one is better for me to learn the distribution of the different modules and I could plot it successfully. Once again, thank you so much!
